File: solutions/Python/remove_duplicates_from_sorted_list_ii_single_pass.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def deleteDuplicates(head: Optional[ListNode]) -> Optional[ListNode]:
    """
    Remove all nodes with duplicate values from sorted linked list in single pass.

    Approach: Use a dummy node and prev pointer. When we find duplicates, skip all nodes
    with that value by advancing current and updating prev.next to point past the group.
    Time: O(n) — single pass through the list
    Space: O(1) — only pointer variables
    """
    if not head:
        return None

    # Create dummy node to handle edge case where head is duplicate
    dummy = ListNode(0)
    dummy.next = head
    prev = dummy
    current = head

    while current:
        # Check if current node is the start of a duplicate group
        if current.next and current.val == current.next.val:
            # Skip all nodes with the same value
            value = current.val
            while current and current.val == value:
                current = current.next
            # Link prev to the first non-duplicate node
            prev.next = current
        else:
            # Current node is unique, keep it
            prev = current
            current = current.next

    return dummy.next


# Test cases
if __name__ == "__main__":
    # Helper function to create linked list from array
    def create_list(arr):
        if not arr:
            return None
        head = ListNode(arr[0])
        current = head
        for val in arr[1:]:
            current.next = ListNode(val)
            current = current.next
        return head

    # Helper function to convert linked list to array
    def list_to_array(head):
        result = []
        current = head
        while current:
            result.append(current.val)
            current = current.next
        return result

    # Test case 1: [1,2,3,3,4,4,5]
    head1 = create_list([1, 2, 3, 3, 4, 4, 5])
    result1 = deleteDuplicates(head1)
    print(list_to_array(result1))  # [1, 2, 5]

    # Test case 2: [1,1,1,2,3]
    head2 = create_list([1, 1, 1, 2, 3])
    result2 = deleteDuplicates(head2)
    print(list_to_array(result2))  # [2, 3]

    # Test case 3: [1,1]
    head3 = create_list([1, 1])
    result3 = deleteDuplicates(head3)
    print(list_to_array(result3))  # []

File: solutions/Rust/remove_duplicates_from_sorted_list_ii_single_pass.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

/**
 * Remove all nodes with duplicate values from sorted linked list in single pass.
 *
 * Approach: Use a dummy node and prev pointer. When we find duplicates, skip all
 * nodes with that value by advancing current and updating prev.next.
 * Time: O(n) — single pass through the list
 * Space: O(1) — only pointer variables
 */
pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    if head.is_none() {
        return None;
    }

    // Create dummy node to handle edge case where head is duplicate
    let mut dummy = Box::new(ListNode::new(0));
    dummy.next = head;
    let mut prev = &mut dummy;
    let mut current = prev.next.take();

    while let Some(mut node) = current {
        current = node.next.take();

        // Check if current node is the start of a duplicate group
        if let Some(next_node) = &current {
            if node.val == next_node.val {
                // Skip all nodes with the same value
                let value = node.val;
                while let Some(n) = current {
                    if n.val != value {
                        current = Some(n);
                        break;
                    }
                    current = n.next;
                }
                // Link prev to the first non-duplicate node
                prev.next = current;
                current = prev.next.take();
                continue;
            }
        }

        // Current node is unique, keep it
        prev.next = Some(node);
        prev = prev.next.as_mut().unwrap();
    }

    dummy.next.take()
}

fn main() {
    // Test case 1: [1,2,3,3,4,4,5]
    let head1 = Some(Box::new(ListNode {
        val: 1,
        next: Some(Box::new(ListNode {
            val: 2,
            next: Some(Box::new(ListNode {
                val: 3,
                next: Some(Box::new(ListNode {
                    val: 3,
                    next: Some(Box::new(ListNode {
                        val: 4,
                        next: Some(Box::new(ListNode {
                            val: 4,
                            next: Some(Box::new(ListNode {
                                val: 5,
                                next: None,
                            })),
                        })),
                    })),
                })),
            })),
        })),
    }));
    println!("Test 1: {:?}", delete_duplicates(head1));  // 1 -> 2 -> 5

    // Test case 2: [1,1,1,2,3]
    let head2 = Some(Box::new(ListNode {
        val: 1,
        next: Some(Box::new(ListNode {
            val: 1,
            next: Some(Box::new(ListNode {
                val: 1,
                next: Some(Box::new(ListNode {
                    val: 2,
                    next: Some(Box::new(ListNode {
                        val: 3,
                        next: None,
                    })),
                })),
            })),
        })),
    }));
    println!("Test 2: {:?}", delete_duplicates(head2));  // 2 -> 3

    // Test case 3: [1,1]
    let head3 = Some(Box::new(ListNode {
        val: 1,
        next: Some(Box::new(ListNode {
            val: 1,
            next: None,
        })),
    }));
    println!("Test 3: {:?}", delete_duplicates(head3));  // null
}
